### ISSUE TITLE
Generate videos from images and text

### ISSUE THREAD

**Issue by blikenoother** [open-source-ideas/open-source-ideas#82]
## Project description

Dynamic video generation based on provided images, text, sound and styling details. Check [this video](https://www.youtube.com/watch?v=PqbYuAMgV7s) where images are coming in the background with zoomin-zoomout effect with text over it.

One can generate a video by passing json config as following:
```json
{
  "audio": "https://s3.amazonaws.com/sounds.wideo.co/6.mp3",
  "fontColor": "#ffffff",
  "fontFamily": "Roboto",
  "fontSize": 24,
  "fontWeight": "200",
  "minDuration": 3000,
  "data": [
    {
      "bgImage": "https://res.cloudinary.com/du8msdgbj/image/upload/c_fit,a_ignore,q_60,fl_lossy,f_auto/v1500545278/marketing/k40niovly9d7uabznqm9.jpg",
      "text": "Medicine overview of TELMA"
    },
    {
      "bgImage": "https://res.cloudinary.com/du8msdgbj/image/upload/c_fit,a_ignore,q_60,fl_lossy,f_auto/v1500545620/marketing/guvqyxyyzuqheg1fy8a1.jpg",
      "text": null
    }
  ],
  "storage": {
    "youtube": {
      "title": "Check this amazing video",
      "description": "Subscribe to our chanel, click on bell icon to get notification",
      "thumbnail": "https://res.cloudinary.com/du8msdgbj/image/upload/c_fit,a_ignore,q_60,fl_lossy,f_auto/v1500545620/marketing/guvqyxyyzuqheg1fy8a1.jpg",
      "privacy": "public"
    },
    "s3": true
  }
}
```
which will generate mp4 video and will upload on youtube, s3 or any other video storage/streaming service

## Relevant Technology

Please suggest how we can achieve this by which technology. I am just mentioning few names which can be used according to me:

- [libopenshot](https://www.openshot.org/libopenshot/)
- fabric.js
- ffmpeg
- YouTube API
- aws S3 API

## Similar Projects

[canvas-sequence](https://github.com/blikenoother/canvas-sequence) HTML5 Canvas + Javascript package (But this is not generating actual video, it's just drawing on canvas with the help of fabric.js)

## Who is this for

- This can be used for dynamic informative videos generation.
- A person who has some numbers and they want to put it with some images and audio and want to generate video

**Comment by straight-shoota** [comment]
I had thought about a similar idea for a simple video composition software. My use case was primarily automatically generating a video from a bunch of images or short video clip (for example event photography).
I've currently solved this problem with a custom script for AfterEffects. It creates a sequence of consecutive images and applies some Ken Burns effect and transitions. It works but it still involves many manual steps. Having a dedicated utility especially for this purpose would be much better

A similar commercial product is https://animoto.com/

A few additional thoughts:

* According to the Unix philosophy, I would keep this focused on the main feature of compositing a video. Uploading to Youtube, AWS or whatever is a different task and there are already great tools for that available.
* I don't think such a tool should be built with web technology (fabric.js). This is likely a task for a sophisticated video editing library (for example [libopenshot](https://www.openshot.org/libopenshot/))
* Advanced feature would synchronize cuts with audio beats
* Maybe it's worth exploring some kind of plugin for [natron](https://natrongithub.github.io/) instead of a stand-alone solution.
* The great benefit an integration in a larger video compositing software is that you can use the automatic generation but also apply some fine tuning. Sometimes you're fine with just hitting a button to generate a nice video. More advanced users might want to change little details to make the result even more awesome, but that can't all be done automatically. Having an easy way to do manual edits, would be great. A scripting interface would also work for this, given the automatic generation is deterministic.

**Comment by blikenoother** [comment]
@straight-shoota yes we should use [libopenshot](https://www.openshot.org/libopenshot/) type of tools but I am not sure they have exposed API to generate video from images. What I can see from their website is they have video editing software and not exposed any API (you can correct me if there is any API available)

**Comment by devAseemSharma** [comment]
On which platform you are targeting this idea? (Web or Mobile) I believe ffmpeg library will be useful in creating the video out of the collection of images and audio.

**Comment by blikenoother** [comment]
@devAseemSharma  I want to build python or nodejs library which accepts json config and return YouTube / aws s3 URL

**Comment by straight-shoota** [comment]
ffmpeg converts video formats. libopenshot can be used for compositing. It provides a feature-rich [API](](http://openshot.org/files/libopenshot/) which should suffice for this use case.

**Comment by schnema123** [comment]
I want to look into this, but libopenshot seems to be a real pain to setup on Windows:

https://openshot.org/files/libopenshot/InstallationGuide.pdf

Even for using the python bindings it seems to be that you need to build the library from scratch (installing all the dependencies, mingw + msys on windows is needed). Does anybody know an easier way to get this library setup?

**Comment by blikenoother** [comment]
@schnema123 you can try this docker: https://hub.docker.com/r/leizongmin/openshot/

**Comment by straight-shoota** [comment]
Or run it on Windows Subsystem for Linux. There you can simply use Linux libraries.

Having an easy-to-use native Windows solution would be great, but that can be dealt with later.

**Comment by blikenoother** [comment]
@straight-shoota @schnema123 @devAseemSharma are you guys in for a short discussion on how roadmap for this project?
Let's schedule an online meeting and set initial thoughts on how to build this

**Comment by FredrikAugust** [comment]
@bikenoother you can use the official slack for this, just hop in and I'll help you set up a channel, team and repo! Link can be found in the Readme :)

**Comment by devAseemSharma** [comment]
@blikenoother definitely, I specialize in Android development, I would be more than happy to contribute as much as I can. Let's get together on Slack for further discussion.

**Comment by FredrikAugust** [comment]
Just hop in @devAseemSharma! There are already some people there you could talk with.
